vm/process/systemVM.go

```diff
@@ -91,10 +111,43 @@ func (s *systemVM) RunSmartContractCall(input *vmcommon.ContractCallInput) (*vmc
 		}, nil
 	}
 
+	lockedGas, err := s.handleAsyncStepGas(input)
+	if err != nil {
+		return &vmcommon.VMOutput{
+			ReturnCode:    vmcommon.UserError,
+			ReturnMessage: err.Error(),
+			GasRemaining:  lockedGas,
+		}, nil
+	}
+
 	returnCode := contract.Execute(input)
 
 	vmOutput := s.systemEI.CreateVMOutput()
 	vmOutput.ReturnCode = returnCode
+	vmOutput.GasRemaining += lockedGas
 
 	return vmOutput, nil
 }
+
+func (s *systemVM) handleAsyncStepGas(input *vmcommon.ContractCallInput) (uint64, error) {
+	if input.CallType != vmcommon.AsynchronousCall {
+		return 0, nil
+	}
+
+	// gasToLock is the amount of gas to set aside for the callback, to avoid it
+	// being used by executing built-in functions; this amount will be restored
+	// to the caller, so that there is sufficient gas for the async callback
+	gasToLock := s.asyncCallStepCost + s.asyncCallbackGasLock
+
+	// gasToDeduct also contains an extra asyncCallStepCost, apart from
+	// gasToLock; asyncCallStepCost will be deducted, but not refunded, just as
+	// Arwen does when executing an async call
+	gasToDeduct := s.asyncCallStepCost + gasToLock
+	if input.GasProvided <= gasToDeduct {
+		return 0, vm.ErrNotEnoughGas
```

Comment:
the lockedgas would be refunded to the user. In this case, although the asyncCall will not happen, the gas for the gasToLock is taken?

Reply:
it would have failed in the sender part first. this is only extra protection.